evm/vm_state.py

```diff
@@ -179,35 +179,27 @@ def get_computation(self, message):
     def apply_transaction(
             self,
             transaction,
-            block,
-            is_stateless=True):
+            block):
         """
         Apply transaction to the given block
 
         :param transaction: the transaction need to be applied
         :param block: the block which the transaction applies on
-        :param is_stateless: if is_stateless, call self.add_transaction to set block
         :type transaction: Transaction
         :type block: Block
-        :type is_stateless: bool
 
         :return: the computation, applied block, and the trie_data
         :rtype: (Computation, Block, dict[bytes, bytes])
         """
-        if is_stateless:
-            # Don't modify the given block
-            block = copy.deepcopy(block)
-            self.set_state_root(block.header.state_root)
-            computation = self.execute_transaction(transaction)
-
-            # Set block.
-            block, trie_data = self.add_transaction(transaction, computation, block)
-            block.header.state_root = self.state_root
-            return computation, block, trie_data
-        else:
-            computation = self.execute_transaction(transaction)
-            block.header.state_root = self.state_root
-            return computation, None, None
+        # Don't modify the given block
+        block = copy.deepcopy(block)
```

Comment:
What do you think about removing this `deepcopy` and instead making `block` immutable and modifying the logic inside of `add_transaction` that creates and returns a new block object.